ships/project-look-and-feel.md

```diff
@@ -0,0 +1,273 @@
+project structure:
+
+```
+.github/
+api/
+client/
+cmd/
+config/
+controllers/
+docs/
+hack/
+pkg/
+test/
+tools/
+vendor/
+.gitignore
+CONTRIBUTING.md
+LICENSE
+go.mod
+go.sum
+main.go
+Makefile
+PROJECT
+README.md
+```
+
+The following directories and files are derived from the Kubebuilder v3 format:
+
+- `api/`: Expected when generating APIs and webhooks
+- `config/`: Always expected
+- `controllers/`: Expected when generating controllers
+- `hack/`: Expected when generating deep copy functions
+- `.gitignore`
+- `LICENSE`
+- `go.mod`
+- `go.sum`
+- `main.go` : Required by kubebuilder
+- `Makefile`
+- `PROJECT`
+
+Keeping this structure ensures that Kubebuilder's command line utilites can be used to simplify
+future development (for example - adding new APIs, controllers, and webhooks).
+
+The following directories are extensions based on our own experiences with Shipwright:
+
+- `.github/`: Configuration for GitHub and GitHub Actions.
+- `client/`: Generated client code. This is optional, and will hold generated go client code.
+- `docs/`: In-tree documentation.
+- `pkg/`: Business logic that can be shared between controllers and/or tested outside of a
+  reconciliation loop.
+- `test/`: Test specifications for integration and end-to-end testing.
+- `tools/`: Additional tooling that is imported with go modules. This an informal [project layout standard](https://github.com/golang-standards/project-layout#tools).
+- `vendor/`: Vendor tree for go code.
+
+### Tooling
+
+Most of the tooling controllers use will be based on the KubeBuilder v3 toolkit.
+These are the baseline tools used to generate code and run integration tests, and should be
+installed locally with Makefile targets:
+
+- Kustomize
+- Controller-gen
+- Envtest
+
+Shipwright controller projects should also standardize on [ko](https://github.com/google/ko) for
+building golang-based container images.
+Ko provides critical features such as reproducible builds and automatic generation of Software
+Bills of Material (SBOMs).
+Ko also simplifies the process of deploying golang containers on Kubernetes, which improves
+contributor experience for testing.
+
+Makefiles should also support a core set of commands, based on what is provided by Kubebuilder:
+
+- `help` - print help for the make commands. This can be generated from comments in the Makefile.
+- `manifests` - generate CRD manifests
+- `generate` - generate go code (DeepCopy, etc.)
+- `fmt` - run `go fmt`
+- `vet` - run `go vet`
+- `test` - run unit tests
+- `test-integration` - run integration tests with Envtest
+- `test-e2e` - run end to end tests against a full Kubernetes cluster
+- `container-build` - build image with ko
+- `container-push` - build and push the image
+- `install` - install CRDs
+- `uninstall` - uninstal CRDs
+- `deploy` - deploy images with Ko
+- `undeploy` - delete deployment
+
+Make targets should also be provided to install build and test tooling:
+
+- `kustomize`
+- `controller-gen`
+- `envtest`
```

Comment:
Will `envtest` really work? Note that we do not build a single controller that we would need to test, but we have the Tekton dependency that also must be present (CRDs) and run (controller). Otherwise many existing integration tests will not run.

Reply:
This IMO is a situation where "words are hard" and definitions are important. The current build "integration" tests require a full cluster and Tekton installed. The tests themselves appear to verify specific feature sets and behaviors, but may not be at first glance "end user" features. The e2e's fit that category, where we exercise the sample build strategies.

I've tended to perceive "integration testing" with Envtest as having a simulated kube environment for our controller. The challenge is that in these tests, we need to simulate the behavior of the Tekton controller.

Reply:
> I've tended to perceive "integration testing" with Envtest as having a simulated kube environment for our controller. The challenge is that in these tests, we need to simulate the behavior of the Tekton controller.

Simulating/mocking the Tekton controller would be not good. Then we potentially miss problems in our code when Tekton changes.

So I am not arguing that envtest could potentially provide reasonable value somewhere. What I primarely want to make sure is that the integration tests that we have today stay there, including their design principle that they have a running Tekton in a running cluster. They are simply an awesome help, easy to write, understand and extend.

At the end it might just be a matter of naming tests (or test suites) (integration vs e2e) and how one defines them. The separation I see today: our integration tests focus on small aspects of the Build and BuildRun controllers that they verify. The e2e test suite does not look at those small pieces but instead verifies that the outcome of the BuildRuns is correct. One can even argue that the only value that e2e tests provide these days is to verify our sample build strategies - which still has a high value.